torch/testing/_internal/common_nn.py

```diff
@@ -3052,16 +3043,16 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
         losses.append(-alpha[-2:].sum().log()[None])
     output = torch.cat(losses, 0)
     if reduction == 'mean':
-        return (output / target_lengths.to(dtype=output.dtype, device=output.device)).mean()
+        output = (output / target_lengths.to(dtype=output.dtype, device=output.device)).mean()
     elif reduction == 'sum':
-        return output.sum()
+        output = output.sum()
     output = output.to(dt)
     return output
```

Comment:
Hm the problem here is that on this line the dtype of `log_probs` is promoted to `float64`, so the `output.to(dt) at the end is necessary for all `reduction` types so that `float32` input --> `float32` output